src/Http/WebUtilities/src/MultipartSectionPipeReader.cs

```diff
@@ -0,0 +1,366 @@
+            }
+
+            //currently there is no handling of a finished read cancellation, so we set false for isCanceled
+            return new ReadResult(_buffer, false, true);
+        }
+
+        public override bool TryRead(out ReadResult result)
+        {
+            if (!_finished)
+            {
+                if (!_pipeReader.TryRead(out result))
+                {
+                    return false;
+                }
+
+                _buffer = result.Buffer;
+                var buffer = result.Buffer;
+                (var didReachEnd, var read) = TryAdvanceToEnd(ref buffer, result.IsCompleted);
+                _buffer = _buffer.Slice(0, read);
+                if (didReachEnd)
+                {
+                    _finished = true;
+                    _pipeReader.AdvanceTo(_buffer.Start, buffer.Start);
+                    _endPosition = buffer.Start;
+                    result = new ReadResult(_buffer, result.IsCanceled, false);
+                    return true;
+                }
+                if (buffer.IsEmpty)
+                {
+                    return true;
+                }
+
+                if (result.IsCompleted)
+                {
+                    throw new IOException("Unexpected end of Stream, the content may have already been read by another component. ");
+                }
+                result = new ReadResult(_buffer, result.IsCanceled, false);
+                return true;
+
+            }
+            if (!_buffer.IsEmpty)
+            {
+                result = new ReadResult(_buffer, false, false);
+                return true;
+            }
+            if (!_metadataSkipped)
+            {
+                if (!_pipeReader.TryRead(out result))
+                {
+                    return false;
+                }
+
+                var sequence = result.Buffer;
+                var isCurrentCompleted = TrySkipMetadata(ref sequence, result.IsCompleted);
+                if (isCurrentCompleted)
+                {
+                    _pipeReader.AdvanceTo(sequence.Start);
+                    _metadataSkipped = true;
+                }
+                else
+                {
+                    _pipeReader.AdvanceTo(sequence.Start, sequence.End);
+                }
+
+                if (result.IsCompleted && !isCurrentCompleted)
+                {
+                    throw new IOException("Unexpected end of Stream, the content may have already been read by another component. ");
+                }
+
+                result = new ReadResult(default, result.IsCanceled, isCurrentCompleted);
+                return true;
+            }
+
+            //currently there is no handling of a finished read cancellation, so we set false for isCanceled
+            result = new ReadResult(_buffer, false, true);
+            return true;
+        }
+
+        private (bool reachedEnd, long consumed) TryAdvanceToEnd(ref ReadOnlySequence<byte> sequence, bool isFinalBlock)
```

Comment:
Since the multi-part format is complex, please comment each method with an example showing the part of the message you're parsing. E.g. `// Scan for partial or full matches of the end-of-section boundary '--boundarymarker--\r\n'`. It helps us reviewers better understand what we're reviewing.